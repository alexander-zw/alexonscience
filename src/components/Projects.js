/**
 * This component contains links to my projects.
 */
import React from "react";
import { Link } from "react-router-dom";
import "../styles/index.css";
import "../styles/Projects.css";
import spacetime_globe from "../images/spacetime_globe/spacetime_globe.png";
import MetaTags from "./subcomponents/MetaTags";

function Projects() {
    return (
        <div>
            <MetaTags
                title="Projects"
                description="Alex's projects"
                keywords="project"
                url="/projects"
            />

            <div className="text-div top-margin bottom-margin">
                <Link className="link" to="/projects/spacetimeglobe">
                    <div className="project-box">
                        <img className="thumbnail" src={spacetime_globe} alt="spacetime globe" />
                        <div className="project-text">
                            <div className="project-title">Spacetime Globe</div>
                            This interactive visualization allows you to play around with Lorentz
                            transformations to really get a feel for how special relativity works.
                            Add events onto the spacetime diagram and see how shifting the reference
                            reference frame affects them!
                        </div>
                    </div>
                </Link>
            </div>
        </div>
    );
}

export default Projects;
