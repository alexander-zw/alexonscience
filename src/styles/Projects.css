.project-box {
    height: 9em;
    margin-left: 4em;
    margin-right: 4em;
    background-color: #f1f1f1;
    border: 0.18em solid grey;
    border-radius: 0.8em;
    display: grid;
    grid-template-columns: 9em auto;
    overflow: hidden;
    transition: box-shadow 0.3s;
}

.project-box:hover {
    box-shadow: 0 0 0.8em #849e9d;
}

.thumbnail {
    width: 9em;
    height: 9em;
    transition: transform 0.3s;
}

.project-box:hover .thumbnail {
    transform: scale(1.2);
}

.project-text {
    padding: 1em 1em 1em 2em;
    border-left: 0.18em solid grey;
    color: black;
    /* Needed to hide the expanding thumbnail. */
    z-index: 5000;
    background-color: #f8f8f8;
}

.project-title {
    font-size: 1.1em;
    font-weight: bold;
    margin-bottom: 0.3em;
    transition: color 0.3s;
}

.project-box:hover .project-title {
    color: #008388;
}

.link {
    text-decoration: none;
}
