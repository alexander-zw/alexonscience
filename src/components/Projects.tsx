import "../styles/index.css";
import "../styles/Projects.css";

import React from "react";
import { Link } from "react-router-dom";

import spacetime_globe from "../images/spacetime_globe/spacetime_globe.png";
import MetaTags from "./subcomponents/MetaTags";

interface ProjectBlockProps {
    path: string;
    title: string;
    image: { src: string; alt: string };
    description: string;
}

const ProjectBlock = ({ path, title, description, image }: ProjectBlockProps) => {
    return (
        <div className="text-div top-margin bottom-margin">
            <Link className="link" to={`/projects/${path}`}>
                <div className="project-box">
                    <img className="thumbnail" src={image.src} alt={image.alt} />
                    <div className="project-text">
                        <div className="project-title">{title}</div>
                        {description}
                    </div>
                </div>
            </Link>
        </div>
    );
};

/**
 * This component contains links to my projects.
 */
function Projects() {
    const projects: ProjectBlockProps[] = [
        {
            path: "spacetimeglobe",
            title: "Spacetime Globe",
            image: { src: spacetime_globe, alt: "spacetime globe" },
            description:
                "This interactive visualization allows you to play around with Lorentz " +
                "transformations to really get a feel for how special relativity works. Add " +
                "events onto the spacetime diagram and see how shifting the reference " +
                "reference frame affects them!",
        },
        {
            path: "explodegifcreator",
            title: "Explode Gif Creator",
            image: { src: spacetime_globe, alt: "explode" },
            description:
                "This tool allows you to turn an image into an exploding gif, which can then " +
                "be turned into an emoji!",
        },
    ];

    const projectBlocks = projects.map((props: ProjectBlockProps) => (
        <ProjectBlock key={props.path} {...props} />
    ));

    return (
        <div>
            <MetaTags path="/projects" />
            {projectBlocks}
        </div>
    );
}

export default Projects;
