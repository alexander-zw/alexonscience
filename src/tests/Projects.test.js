/**
 * Test for the Projects.js component.
 */
import React from "react";
import { BrowserRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import Projects from "../components/Projects";

it("renders correctly", () => {
    const tree = renderer
        .create(
            <BrowserRouter>
                <Projects />
            </BrowserRouter>,
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
