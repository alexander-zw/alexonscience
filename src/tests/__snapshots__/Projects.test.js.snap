// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="text-div top-margin bottom-margin"
  >
    <a
      className="link"
      href="/projects/spacetimeglobe"
      onClick={[Function]}
    >
      <div
        className="project-box"
      >
        <img
          alt="spacetime globe"
          className="thumbnail"
          src="spacetime_globe.png"
        />
        <div
          className="project-text"
        >
          <div
            className="project-title"
          >
            Spacetime Globe
          </div>
          This interactive visualization allows you to play around with Lorentz transformations to really get a feel for how special relativity works. Add events onto the spacetime diagram and see how shifting the reference reference frame affects them!
        </div>
      </div>
    </a>
  </div>
</div>
`;
